src/cluster_resources.rs

```diff
@@ -0,0 +1,404 @@
+    /// * `app.kubernetes.io/managed-by = <app_name>-operator`
+    /// * `app.kubernetes.io/name = <app_name>`
+    ///
+    /// If the patched resource does not contain a UID then an `Error::MissingObjectKey` is
+    /// returned.
+    pub async fn add<T: ClusterResource>(
+        &mut self,
+        client: &Client,
+        resource: &T,
+    ) -> OperatorResult<T> {
+        ClusterResources::check_label(resource.labels(), APP_INSTANCE_LABEL, &self.app_instance)?;
+        ClusterResources::check_label(resource.labels(), APP_MANAGED_BY_LABEL, &self.manager)?;
+        ClusterResources::check_label(resource.labels(), APP_NAME_LABEL, &self.app_name)?;
+
+        let patched_resource = client
+            .apply_patch(&self.manager, resource, resource)
+            .await?;
+
+        let resource_id = patched_resource.uid().ok_or(Error::MissingObjectKey {
+            key: "metadata/uid",
+        })?;
+
+        self.resource_ids.insert(resource_id);
+
+        Ok(patched_resource)
+    }
+
+    /// Checks that the given `labels` contain the given `label` with the given `expected_content`.
+    ///
+    /// # Arguments
+    ///
+    /// * `labels` - The labels to check
+    /// * `label` - The expected label
+    /// * `expected_content` - The expected content of the label
+    ///
+    /// # Errors
+    ///
+    /// If `labels` does not contain `label` then an `Error::MissingLabel` is returned.
+    ///
+    /// If `labels` contains the given `label` but not with the `expected_content` then an
+    /// `Error::UnexpectedLabelContent` is returned
+    fn check_label(
+        labels: &BTreeMap<String, String>,
+        label: &'static str,
+        expected_content: &str,
+    ) -> OperatorResult<()> {
+        if let Some(actual_content) = labels.get(label) {
+            if expected_content == actual_content {
+                Ok(())
+            } else {
+                Err(Error::UnexpectedLabelContent {
+                    label,
+                    expected_content: expected_content.into(),
+                    actual_content: actual_content.into(),
+                })
+            }
+        } else {
+            Err(Error::MissingLabel { label })
+        }
+    }
+
+    /// Finalizes the cluster creation.
+    ///
+    /// The orphaned resources of all kinds of resources which implement the [`ClusterResource`]
+    /// trait, are deleted. A resource is seen as orphaned if it is labelled as if it belongs to
+    /// this cluster instance but was not added to the cluster resources before.
+    ///
+    /// The following resource labels are compared:
+    /// * `app.kubernetes.io/instance`
+    /// * `app.kubernetes.io/managed-by`
+    /// * `app.kubernetes.io/name`
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    pub async fn finalize(self, client: &Client) -> OperatorResult<()> {
+        self.delete_orphaned_resources::<Service>(client).await?;
```

Comment:
There's no need to wait for each deletion sequentially, `tokio::join!` can be used to execute deletion of all the resources concurrently.